data/blog/how-open-pr-comments-work.md

```diff
@@ -0,0 +1,124 @@
+
+#### Extract functions from the file patch
+
+1.  Depending on the language of the file, we fetch the appropriate parser and apply it to the file patch (the git diff for the file).
+2.  This applies a regex to the whole file, looking for git hunk headers (e.g. `@@ -188,9 +188,7 @@ def __init__():` for Python) that indicate a function in the language, and that the code modified in the section below it belongs to that function. This is more or less correct in finding functions being modified in a PR, unless the function is super short or the lines modified are near the top of the function (this is a limitation of git).
+3.  For Python, the regex only looks for `def {function_name}.` For Javascript/Typescript, there are more ways to initialize a function (function declaration, arrow function, function expression, etc), so there are multiple regexes to find function names.
+
+#### Snuba query to fetch top 5 issues by count
+
+1.  We first fetch the first 10k unresolved issues for the projects found via reverse codemapping through Postgres, ordered by times seen. This is to prevent overloading the Snuba query if we pass in too many issue ids.
+2.  Next, we have a complicated looking Snuba query that does the following:
+    - Subquery to fetch the count of events for each issue id
+    - Query that filters on the subquery to 1) squash issue ids with the same title and culprit, 2) look for unhandled events that have the filename+function combo for any of the file’s function within the first 4 frames of the stacktrace, and 3) return the top 5 issues with the greatest count of events
+3.  There is also some different logic being inserted depending on the language of the file. For instance, for Javascript/Typescript we also look for events that have `{any classname}.function_name` inside the file in addition to just `function_name` inside the file because of how Javascript/Typescript events are stored.
+
+### Create comment
+
+Then for each file, we make a little table with the function and issue information. All the tables besides the one for the first file are hidden in a toggle. Each file type may also have a slightly different formatting template. In Javascript/Typescript, we want to show the `Affected Users` because it’s more important for frontend. Meanwhile, Python usually always has 0 affected users so it’s not shown.
+
+![Javascript open PR comment](/images/open-pr-comments/open-pr-comment.png)
+
+## A note on language parsers
+
+We support different languages in open PR comments. However, they also require different methods to extract functions from the diff, different handling for the ways events with those functions might be stored in Snuba, and how to format the comment table for each file type.
```

Comment:
Oh I guess the answer is here. Should we mention which ones or link to the docs so it evolves with time?

Reply:
yeah i'll link to docs